lib/Sema/ConstraintPrinter.cpp

```diff
@@ -0,0 +1,363 @@
+
+    /// @brief Print the constraint kind as a string.
+    /// @param kind The constraint kind to print.
+    /// @param os The output stream to print to.
+    void printConstraintKind(ConstraintKind kind, llvm::raw_ostream &os);
+
+    /// @brief Format constraint details as a string.
+    /// @param constraint The constraint to format.
+    /// @return A formatted string with constraint details.
+    std::string formatConstraintDetails(Constraint const *constraint);
+
+    /// @brief Format a type as a string.
+    /// @param type The type to format.
+    /// @return A formatted string representation of the type.
+    std::string formatType(glu::types::Ty type);
+
+    /// @brief Format a function declaration as a string.
+    /// @param decl The function declaration to format.
+    /// @return A formatted string representation of the function.
+    std::string formatFunctionDecl(glu::ast::FunctionDecl const *decl);
+
+    /// @brief Print statement context before constraints.
+    /// @param stmt The statement to print context for.
+    /// @param os The output stream to print to.
+    void printStatementContext(
+        glu::sema::ConstraintSystem &system, llvm::raw_ostream &os
+    );
+
+private:
+    /// @brief Recursively print a constraint and its nested constraints.
+    /// @param constraint The constraint to print.
+    /// @param os The output stream to print to.
+    /// @param indent The indentation level for formatting.
+    void printConstraintRecursive(
+        Constraint const *constraint, llvm::raw_ostream &os, unsigned indent
+    );
+};
+
+void Constraint::print(llvm::raw_ostream &os, unsigned indent) const
+{
+    os.indent(indent);
+
+    // Use a single ConstraintPrinter instance
+    ConstraintPrinter printer;
+
+    // Print constraint kind name
+    printer.printConstraintKind(getKind(), os);
+    os << ": ";
+
+    // Print constraint details using the structured format
+    std::string details = printer.formatConstraintDetails(this);
+    os << details;
+
+    os << "\n";
+}
+
+void ConstraintPrinter::print(ConstraintSystem &system, llvm::raw_ostream &os)
+{
+
+    printStatementContext(system, os);
+
+    auto const &constraints = system.getConstraints();
+
+    os << "======== ConstraintSystem with " << constraints.size()
+       << " constraint(s) ========\n";
+
+    for (size_t i = 0; i < constraints.size(); ++i) {
+        os << "[" << i << "] ";
+        printConstraintRecursive(constraints[i], os, 0);
+    }
+
+    if (constraints.empty()) {
+        os << "  (no constraints)\n";
+    } else {
+        os << "============================================="
+              "==========\n";
+    }
+}
+
+void ConstraintPrinter::printConstraintRecursive(
+    Constraint const *constraint, llvm::raw_ostream &os, unsigned indent
+)
+{
+    if (!constraint) {
+        os.indent(indent);
+        os << "<null constraint>\n";
+        return;
+    }
+
+    constraint->print(os, indent);
```

Comment:
don't call this here, you lose `_typePrinter`
keep the implementation within this class